cmd/fsck.go

```diff
@@ -51,172 +51,200 @@ $ juicefs fsck redis://localhost --path /d1/d2 --repair
-			logger.Fatalf("File path should be the absolute path within JuiceFS")
-		}
-		return m.Check(c, p, ctx.Bool("repair"), ctx.Bool("recursive"), ctx.Bool("sync-dir-stat"))
-	}
 
-	chunkConf := chunk.Config{
-		BlockSize:  format.BlockSize * 1024,
-		Compress:   format.Compression,
-		GetTimeout: time.Second * 60,
-		PutTimeout: time.Second * 60,
-		MaxUpload:  20,
-		BufferSize: 300 << 20,
-		CacheDir:   "memory",
+	if ctx.Bool("sync-dir-stat") && !format.DirStats {
+		logger.Warn("dir stats is disabled, flag '--sync-dir-stat' will be ignored")
 	}
 
-	blob, err := createStorage(*format)
-	if err != nil {
-		logger.Fatalf("object storage: %s", err)
-	}
-	logger.Infof("Data use %s", blob)
-	blob = object.WithPrefix(blob, "chunks/")
-	objs, err := osync.ListAll(blob, "", "", "", true)
-	if err != nil {
-		logger.Fatalf("list all blocks: %s", err)
+	threads := ctx.Uint("threads")
+	if threads == 0 {
+		threads = 20
+		logger.Warnf("threads number set to %d", threads)
 	}
 
-	// Find all blocks in object storage
+	var c = meta.NewContext(0, 0, []uint32{0})
 	progress := utils.NewProgress(false)
-	blockDSpin := progress.AddDoubleSpinner("Found blocks")
-	var blocks = make(map[string]int64)
-	for obj := range objs {
-		if obj == nil {
-			break // failed listing
-		}
-		if obj.IsDir() {
-			continue
+
+	var metaChecker func(meta.Ino, string, *meta.Attr) error
+	if flag&checkMetaFlag != 0 {
+		metaChecker = m.GetMetaChecker(c, ctx.Bool("repair"), ctx.Bool("sync-dir-stat"))
+	}
+
+	var dataChecker func(meta.Ino, []meta.Slice) error
+	brokens := make(map[meta.Ino]string)
+	if flag&checkDataFlag != 0 {
+		sliceCBar := progress.AddCountBar("Scanned slices", 0)
+		sliceBSpin := progress.AddByteSpinner("Scanned slices")
+		lostDSpin := progress.AddDoubleSpinner("Lost blocks")
+		defer func() {
+			if progress.Quiet {
+				logger.Infof("Used by %d slices (%d bytes)", sliceCBar.Current(), sliceBSpin.Current())
+			}
+			if lc, lb := lostDSpin.Current(); lc > 0 {
+				msg := fmt.Sprintf("%d objects are lost (%d bytes), %d broken files:\n", lc, lb, len(brokens))
+				msg += fmt.Sprintf("%13s: PATH\n", "INODE")
+				var fileList []string
+				for i, p := range brokens {
+					fileList = append(fileList, fmt.Sprintf("%13d: %s", i, p))
+				}
+				sort.Strings(fileList)
+				msg += strings.Join(fileList, "\n")
+				logger.Fatal(msg)
+			}
+		}()
+
+		chunkConf := chunk.Config{
+			BlockSize:  format.BlockSize * 1024,
+			Compress:   format.Compression,
+			GetTimeout: time.Second * 60,
+			PutTimeout: time.Second * 60,
+			MaxUpload:  20,
+			BufferSize: 300 << 20,
+			CacheDir:   "memory",
 		}
 
-		logger.Debugf("found block %s", obj.Key())
-		parts := strings.Split(obj.Key(), "/")
-		if len(parts) != 3 {
-			continue
+		blob, err := createStorage(*format)
+		if err != nil {
+			logger.Fatalf("object storage: %s", err)
+		}
+		logger.Infof("Data use %s", blob)
+		blob = object.WithPrefix(blob, "chunks/")
+		objs, err := osync.ListAll(blob, "", "", "", true)
```

Comment:
it's slow to list all objects when we only need to check a single file.
We should use Head to check each object instead.

Reply:
fixed